packages/hardhat-ledger/src/provider.ts

```diff
@@ -0,0 +1,445 @@
+
+      await this._requireControlledInit(address);
+
+      if (address !== undefined) {
+        if (data === undefined) {
+          throw new HardhatError(ERRORS.NETWORK.ETHSIGN_MISSING_DATA_PARAM);
+        }
+
+        const path = await this._derivePath(address);
+        const signature = await this._withConfirmation(() =>
+          this.eth.signPersonalMessage(path, data.toString("hex"))
+        );
+        return this._toRpcSig(signature);
+      }
+    }
+  }
+
+  private async _personalSign(params: any[]): Promise<unknown> {
+    if (params.length > 0) {
+      const [data, address] = validateParams(params, rpcData, rpcAddress);
+
+      await this._requireControlledInit(address);
+
+      if (data !== undefined) {
+        if (address === undefined) {
+          throw new HardhatError(
+            ERRORS.NETWORK.PERSONALSIGN_MISSING_ADDRESS_PARAM
+          );
+        }
+
+        const path = await this._derivePath(address);
+        const signature = await this._withConfirmation(() =>
+          this.eth.signPersonalMessage(path, data.toString("hex"))
+        );
+        return this._toRpcSig(signature);
+      }
+    }
+  }
+
+  private async _ethSignTypedDataV4(params: any[]): Promise<unknown> {
+    const [address, data] = validateParams(params, rpcAddress, t.any as any);
+
+    await this._requireControlledInit(address);
+
+    if (data === undefined) {
+      throw new HardhatError(ERRORS.NETWORK.ETHSIGN_MISSING_DATA_PARAM);
+    }
+
+    let typedMessage: EIP712Message;
+    try {
+      typedMessage = typeof data === "string" ? JSON.parse(data) : data;
+
+      if (!isEIP712Message(typedMessage)) {
+        throw new HardhatError(
+          ERRORS.NETWORK.ETHSIGN_TYPED_DATA_V4_INVALID_DATA_PARAM
+        );
+      }
+    } catch {
+      throw new HardhatError(
+        ERRORS.NETWORK.ETHSIGN_TYPED_DATA_V4_INVALID_DATA_PARAM
+      );
+    }
+
+    const { types, domain, message, primaryType } = typedMessage;
+    const { EIP712Domain: _, ...structTypes } = types;
+
+    const path = await this._derivePath(address);
+    const signature = await this._withConfirmation(async () => {
+      try {
+        return await this.eth.signEIP712Message(path, typedMessage);
+      } catch (error) {
+        return this.eth.signEIP712HashedMessage(
+          path,
+          ethers.TypedDataEncoder.hashDomain(domain),
+          ethers.TypedDataEncoder.hashStruct(primaryType, structTypes, message)
+        );
+      }
+    });
+
+    return this._toRpcSig(signature);
+  }
+
+  private async _ethSendTransaction(params: any[]): Promise<unknown> {
+    const [txRequest] = validateParams(params, rpcTransactionRequest);
+
+    await this._requireControlledInit(txRequest.from);
+
+    if (txRequest.gas === undefined) {
+      throw new HardhatError(ERRORS.NETWORK.MISSING_TX_PARAM_TO_SIGN_LOCALLY, {
+        param: "gas",
+      });
+    }
+
+    if (txRequest.from === undefined) {
```

Comment:
This check doesnt' seem necessary. `validateParams` already validates that it's not undefined, and anyway the `this._requireControlledInit` is already assuming that it's not undefined.

Reply:
Right you are, good catch!